### DISCUSSION TITLE
[Deployment] <Getting error while deploying transparent proxy contract>

### DISCUSSION THREAD

**Discussion by FurkanSezal** [zkSync-Community-Hub/zksync-developers#301] [Deployments] [answered]
### Environment

Testnet

### zkSolc Version

latest

### zksync-web3 Version

^0.14.3

### Hardhat.config.ts

```javascript
import { HardhatUserConfig } from "hardhat/config";
import "@nomicfoundation/hardhat-toolbox";
import "@nomiclabs/hardhat-ethers";
import "@openzeppelin/hardhat-upgrades";
import "hardhat-deploy";
import "@nomiclabs/hardhat-ethers";
import "solidity-coverage";
import "@matterlabs/hardhat-zksync-deploy";
import "@matterlabs/hardhat-zksync-solc";
import "@matterlabs/hardhat-zksync-upgradable";

import dotenv from "dotenv";

dotenv.config();

const PRIVATE_KEY = process.env.PRIVATE_KEY;
if (!PRIVATE_KEY) {
  throw Error("process.env.PRIVATE_KEY IS REQUIRED");
}
const SEPOLIA_RPC_URL = process.env.SEPOLIA_RPC_URL;
if (!SEPOLIA_RPC_URL) {
  throw Error("process.env.SEPOLIA_RPC_URL IS REQUIRED");
}

const config: HardhatUserConfig = {
  solidity: {
    version: "0.8.18",
    settings: {
      optimizer: {
        enabled: true,
        runs: 200,
      },
    },
  },
  zksolc: {
    version: "latest",
    compilerSource: "binary",
    settings: {},
  },
  defaultNetwork: "zkTestnet",
  namedAccounts: {
    deployer: {
      default: 0,
    },
  },
  networks: {
    hardhat: {
      allowUnlimitedContractSize: true,
    },

    zkTestnet: {
      url: "https://sepolia.era.zksync.dev",
      ethNetwork: SEPOLIA_RPC_URL,
      zksync: true,
      chainId: 300,
    },
  },


};

export default config;
```

### Deployment Script (WITHOUT PRIVATE KEY)

```javascript
import { HardhatRuntimeEnvironment } from "hardhat/types";
import { ABI, DeployFunction } from "hardhat-deploy/types";
import { PAYMENT_TOKENS } from "../constants";

import { utils, Wallet, Provider } from "zksync-web3";
import { Deployer } from "@matterlabs/hardhat-zksync-deploy";

const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
  const contractName = "MarketPlace";
  console.log("Deploying " + contractName + "...");

  const PRIVATE_KEY: string = process.env.PRIVATE_KEY!;

  const zkWallet = new Wallet(PRIVATE_KEY);

  const deployer = new Deployer(hre, zkWallet);

  console.log("Deployer address: ", deployer.ethWallet.address);

  const network = await hre.ethers.provider.getNetwork();

  const chainId = network.chainId.toString();

  if (!Object.keys(PAYMENT_TOKENS).includes(chainId) && chainId !== "31337") {
    throw new Error(`No payment tokens for chainId ${chainId}`);
  }

  let paymentTokens: string[] = PAYMENT_TOKENS[chainId as keyof typeof PAYMENT_TOKENS];

  const contract = await deployer.loadArtifact(contractName);

  const contractDeployed = await hre.zkUpgrades.deployProxy(
    deployer.zkWallet,
    contract,
    [
      paymentTokens,
      "0x02dF0d9BA51Cb7E62f28fa2884c0AFaa0D84B3Cd",
      deployer.ethWallet.address, // owner
    ],
    { initializer: "initialize" }
  );

  await contractDeployed.deployed();
  console.log(contractName + " deployed to:", contractDeployed.address);
};
module.exports = func;

module.exports.tags = ["Deploy"];
```

### Package.json

```json
{
  "name": "hardhat-project",
  "dependencies": {
    "@ethersproject/providers": "^5.7.2",
    "@matterlabs/hardhat-zksync-deploy": "^0.6.3",
    "@matterlabs/hardhat-zksync-solc": "^0.4.1",
    "@matterlabs/hardhat-zksync-upgradable": "^0.1.2",
    "@nomicfoundation/hardhat-chai-matchers": "^1.0.6",
    "@nomicfoundation/hardhat-network-helpers": "^1.0.8",
    "@nomicfoundation/hardhat-toolbox": "^2.0.2",
    "@nomiclabs/hardhat-ethers": "^2.0.3",
    "@nomiclabs/hardhat-etherscan": "^3.1.7",
    "@openzeppelin/contracts": "^4.8.2",
    "@openzeppelin/contracts-upgradeable": "^4.8.2",
    "@openzeppelin/hardhat-upgrades": "^1.22.1",
    "@openzeppelin/upgrades": "^2.8.0",
    "@openzeppelin/upgrades-core": "1.27.3",
    "@typechain/ethers-v5": "^10.2.0",
    "@typechain/hardhat": "^6.1.5",
    "@types/chai": "^4.3.4",
    "@types/mocha": "^10.0.1",
    "@types/node": "^18.15.3",
    "chai": "^4.3.7",
    "dotenv": "^16.0.3",
    "ethers": "^5.7.2",
    "hardhat": "^2.17.0",
    "hardhat-deploy": "^0.11.25",
    "hardhat-deploy-ethers": "^0.3.0-beta.13",
    "hardhat-gas-reporter": "^1.0.9",
    "moment": "^2.29.4",
    "solidity-coverage": "^0.8.2",
    "solmate": "^6.7.0",
    "ts-node": "^10.9.1",
    "typechain": "^8.1.1",
    "typescript": "^5.0.2",
    "web3-utils": "^1.8.2",
    "zksync-web3": "^0.14.3"
  },
  "devDependencies": {
    "fs": "^0.0.1-security",
    "path": "^0.12.7"
  }
}
```

### Contract Code

```solidity
an nft marketplace contract
```

### Does this work on other EVMs? (If yes, please list at least 1 of them)

yes. bnb,ethereum,polygon, zksync mainnet 

### Description of What Your Contract Does

Its an nft marketplace 

### Additional Details

This contract was on zksync goerli testnet. Now i want to migrate to sepolia. Its deploy Implementation and admin contract but not proxy.But i am getting this error. 
By the way this already on zksync mainnet as well. 

Deploying MarketPlace...
Deployer address:  0xEaC9eDFE37fA378E8795253d292e6393d29aBCa2
Implementation contract was deployed to 0xc3Ab54d545A496d684A57FC44533628f1e641307
Admin was deployed to 0x963Bb5caB941165f4353Db9Fbc9D2b19ce4070F9
An unexpected error occurred:
Error: ERROR processing /Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/deploy/2-Bit5.ts:
Error: cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (error={"reason":"execution reverted: Address: low-level delegate call failed","code":"UNPREDICTABLE_GAS_LIMIT","method":"estimateGas","transaction":{"from":"0xEaC9eDFE37fA378E8795253d292e6393d29aBCa2","to":"0x0000000000000000000000000000000000008006","data":""],"gasPerPubdata":50000},"from":"0xEaC9eDFE37fA378E8795253d292e6393d29aBCa2","nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.7.0)
    at Logger.makeError (/Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
    at Logger.throwError (/Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
    at /Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/@ethersproject/abstract-signer/src.ts/index.ts:301:31
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Promise.all (index 6)
    at DeploymentsManager.executeDeployScripts (/Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/hardhat-deploy/src/DeploymentsManager.ts:1225:19)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async DeploymentsManager.runDeploy (/Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/hardhat-deploy/src/DeploymentsManager.ts:1055:5)
    at async SimpleTaskDefinition.action (/Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/hardhat-deploy/src/index.ts:438:5)
    at async Environment._runTaskDefinition (/Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:333:14)
    at async Environment.run (/Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:166:14)
    at async SimpleTaskDefinition.action (/Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/hardhat-deploy/src/index.ts:584:32)
    at async Environment._runTaskDefinition (/Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:333:14)
    at async Environment.run (/Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:166:14)
    at async SimpleTaskDefinition.action (/Users/furkansezal/Desktop/Bit5ZkSyncEraContracts/node_modules/hardhat-deploy/src/index.ts:669:5)
error Command failed with exit code 1.

**Comment by MexicanAce** [comment]
Hi @FurkanSezal 👋 !

It's a little tough to tell from the information above what could be causing this deployment error just for Sepolia Testnet, is there any chance you could share your smart contracts with us?

In the meantime, please recompile the contracts and try again, perhaps this was an issue with the cached local contracts.

**Reply by FurkanSezal** [reply]
Yes i already tried that. Also i tried to upgrade all the packages with ethers 6 as well. I am getting same error while deploying the proxy contract. 
The contract code is private code. I can share in private is it okey with you?

**Reply by MexicanAce** [reply]
@FurkanSezal that works.

**Reply by AlbionaHoti** [reply]
Hi @FurkanSezal, how's the progress with your project? Are you still having the same issue? If yes, it would help us if you could share your code so we can debug it.

**Reply by FurkanSezal** [reply]
@AlbionaHoti hi, I have still same issue. Can i share with you the project?

**Reply by bxpana** [reply]
@FurkanSezal wanted to check on this and see if you were able to get support on this. If not, let me know and you can add me to the project to continue with help.

**Reply by FurkanSezal** [reply]
@bxpana Hello, yes i couldnt solve the issue. I just send you our contract code via email could you check please

**Reply by bxpana** [reply]
Got your email and will take some time to review the code and get back to you 😁

**Reply by dutterbutter** [reply]
@FurkanSezal it seems you do not have the verify plugin installed. My suggestions:
- Update your dependencies to make use of `@matter-labs/hardhat-zksync` and target the ethers v5 version at v0.1.0 [here](https://github.com/matter-labs/hardhat-zksync/releases/tag/%40matterlabs%2Fhardhat-zksync-v0.1.0). As it seems you are using ethers v5 in your project.
- `@matter-labs/hardhat-zksync` provides all the plugins you need so you can remove  
     - "@matterlabs/hardhat-zksync-deploy": "^0.6.3",
    - "@matterlabs/hardhat-zksync-solc": "^0.4.1",
    - "@matterlabs/hardhat-zksync-upgradable": "^0.1.2",
- Remove `zksync-web3` dependency in favour of `zksync-ethers` at v0.5.7 [here](https://github.com/zksync-sdk/zksync-ethers/releases/tag/v5.7.0)

Then once upgradeable contract is deployed, you can verify it with:

`yarn hardhat verify <PROXY-ADDRESS>`

**Reply by FurkanSezal** [reply]
@dutterbutter I dont want to verify it yet. I tried upgrade npm packages. It didnt work. I am getting same error while trying to deploy. Its deploys implementation and proxy admin and while deploying proxy getting this error : 
error={"reason":"execution reverted: Address: low-level delegate call failed"}
did you try to deploy to zksync sepolia?

**Reply by bxpana** [reply]
@FurkanSezal making some progress here and will hopefully have more info for you next week. Appreciate your patience.

**Reply by bxpana** [reply, marked as answer]
There were quite a few changes I made to be able to deploy on Sepolia. Would it be easier if you let me create a PR for the repo and share it so you guys could see the differences or should I create a private PR and invite you to it?

**Reply by bxpana** [reply]
@FurkanSezal submitted the PR and we can continue the conversation in the comments of the PR

**Comment by jablonskibrittany** [comment]
The error "cannot estimate gas; transaction may fail or may require manual gas limit" typically occurs when the contract deployment fails to estimate the gas limit required for the transaction. This can happen due to various reasons, such as incorrect contract initialization or insufficient gas limit provided.

To resolve this issue, you can try increasing the gas limit for the transaction manually or ensuring that the contract initialization parameters are correct. Additionally, verifying that the network settings and provider URLs are configured correctly might help resolve any connectivity issues that could cause transact

**Comment by berronesjosh** [comment]
# This error typically occurs when there's an issue with estimating gas limits for a transaction. It can happen due to various reasons such as insufficient funds, incorrect contract initialization, or contract deployment on a different network. To resolve this, ensure that the contract deployment parameters are correct, the account has sufficient funds, and the deployment is targeting the correct network (in this case, zkSync Testnet).

**Comment by renemilnev** [comment]
This error suggests that there might be an issue with estimating gas or executing a low-level delegate call during deployment. Further investigation into the specific contract implementation and deployment settings is needed to resolve this issue.
